### PR TITLE
Unified Storage: Fix Create, Update and Delete wrt Resource Versions

### PR DESCRIPTION
**Pull request by diegommm**
**What is this feature?**

1. Rewrite the Create, Update and Delete methods of Unified Storage to adhere to Proposal 1 of Consistent Resource Version Design Doc.
2. Fix a lot of things that make it too difficult to do progress, only flagging with a comment when the fix can be postponed.
3. Fix Unified Storage to allow running it with the API Server, currently on `main` only the GRPC server works (needed it to correctly test).

**Why do we need this feature?**

This enables a consistent preview of Consisten Resource Version Proposal 1 write path. This works for SQLite, MySQL, Vitess and PostgreSQL as backends. CockroachDB will also work but will be highly penalized due to its Default Isolation level being Serializable. When the read path is also covered in a future PR, CockroachDB will work better since in the read path we also change the Isolation Level to Read Committed, which is supported by all the relevant backends.

**Who is this feature for?**

Everyone.

**Which issue(s) does this PR fix?**:

None

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/store/entity/sqlstash/sql_storage_server.go [hunk 1]
```diff
@@ -75,13 +80,27 @@ type sqlEntityServer struct {
 	dialect     migrator.Dialect
 	snowflake   *snowflake.Node
 	broadcaster Broadcaster[*entity.EntityWatchResponse]
-	ctx         context.Context
+	ctx         context.Context // TODO: remove
 	cancel      context.CancelFunc
 	stream      chan *entity.EntityWatchResponse
 	tracer      tracing.Tracer
+
+	once    sync.Once
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/sql_storage_server.go:88]: Fixing another annoying data race!

File: pkg/services/store/entity/sqlstash/sqltemplate/args.go [hunk 2]
```diff
@@ -13,6 +24,47 @@ func (a *Args) Arg(x any) string {
...
+//		*sqltemplate.SQLTemplate
+//		IDs []int64
+//	}
+//
+// Example usage in a SQL template:
+//
+//	DELETE FROM {{ .Ident "mytab" }}
+//		WHERE id IN ( {{ argList . .IDs }} )
+//	;
+func (a *Args) ArgList(slice reflect.Value) (string, error) {
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/sqltemplate/args.go:41]: Tests in another PR, too long already

File: pkg/services/store/entity/sqlstash/sqltemplate/sqltemplate.go [hunk 3]
```diff
@@ -33,3 +51,38 @@ func Execute(t *template.Template, data any) (string, error) {
...
+// it makes the following assumptions, which are also good practices for writing
+// your SQL templates:
+//  1. There are no SQL comments. Consider adding your comments as template
+//     comments instead (i.e. "{{/* this is a template comment */}}").
+//  2. There are no multiline strings, and strings do not contain consecutive
+//     spaces. Code looking like this is already a smell. Avoid string literals,
+//     pass them as arguments so they can be appropriately escaped by the
+//     corresponding driver. And identifiers with white space should be avoided
+//     in all cases as well.
+func FormatSQL(q string) string {
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/sqltemplate/sqltemplate.go:67]: Tests in another PR, too long already

File: pkg/services/store/entity/sqlstash/testdata/simple.jsonc [hunk 4]
```diff
@@ -3,16 +3,16 @@
 //  Frame[0] 
 //  Name: 
 //  Dimensions: 7 Fields by 4 Rows
-//  +----------------+----------------+----------------+---------------+---------------+---------------+---------------------------------------------------------------------------------------------------------+
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/testdata/simple.jsonc:6]: This is autogenerated.

File: pkg/services/store/entity/sqlstash/update.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
+package sqlstash
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/update.go:1]: Tests in another PR, too long already

File: pkg/services/store/entity/sqlstash/utils.go [hunk 6]
```diff
@@ -1,15 +1,172 @@
 package sqlstash
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/utils.go:1]: Tests in another PR, too long already

File: pkg/services/store/entity/sqlstash/create.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
+package sqlstash
```
Comment (diegommm) [inline pkg/services/store/entity/sqlstash/create.go:1]: Tests in another PR, too long already

File: pkg/services/apiserver/storage/entity/test/watch_test.go [hunk 8]
```diff
@@ -153,6 +153,7 @@ func TestIntegrationWatch(t *testing.T) {
 	if testing.Short() {
 		t.Skip("skipping integration test")
 	}
+	t.Skip("In maintenance")
```
Comment (diegommm) [inline pkg/services/apiserver/storage/entity/test/watch_test.go:156]: We disable k8s integration tests until we have also correctly covered the read path as well.

File: pkg/services/store/entity/db/dbimpl/dbEngine.go [hunk 9]
```diff
@@ -1,75 +1,98 @@
...
+	"github.com/grafana/grafana/pkg/infra/tracing"
+	"github.com/grafana/grafana/pkg/services/store/entity/db"
 )
 
-func getEngineMySQL(cfgSection *setting.DynamicSection, tracer tracing.Tracer) (*xorm.Engine, error) {
-	dbHost := cfgSection.Key("db_host").MustString("")
-	dbName := cfgSection.Key("db_name").MustString("")
-	dbUser := cfgSection.Key("db_user").MustString("")
-	dbPass := cfgSection.Key("db_pass").MustString("")
+func getEngineMySQL(getter *sectionGetter, _ tracing.Tracer) (*xorm.Engine, error) {
```
Comment (chaudyg) [inline pkg/services/store/entity/db/dbimpl/dbEngine.go:16]: There are a few minor differences between this implementation and the previous one:
- sql_mode: @@SESSION.sql_mode": "TRADITIONAL,ANSI"
- Loc: time.UTC
- New config flags :  db_server_pub_key / db_tls_config_name
- CheckConnLiveness = true

I suggest we keep this fully backward compatible for now.

File: pkg/services/store/entity/db/dbimpl/dbEngine.go [hunk 10]
```diff
@@ -1,75 +1,98 @@
...
 
 	return engine, nil
 }
 
-func getEnginePostgres(cfgSection *setting.DynamicSection, tracer tracing.Tracer) (*xorm.Engine, error) {
-	dbHost := cfgSection.Key("db_host").MustString("")
-	dbName := cfgSection.Key("db_name").MustString("")
-	dbUser := cfgSection.Key("db_user").MustString("")
-	dbPass := cfgSection.Key("db_pass").MustString("")
+func getEnginePostgres(getter *sectionGetter, _ tracing.Tracer) (*xorm.Engine, error) {
```
Comment (chaudyg) [inline pkg/services/store/entity/db/dbimpl/dbEngine.go:56]: changing this makes me a bit nervious. Could we keep this in a different PR ?

File: pkg/services/store/entity/sqlstash/data/entity_history.sql [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+            AND {{ .Ident "guid" }} = {{ .Arg .Query.GUID }}
+
+        {{ else }}
+            AND {{ .Ident "group" }}    = {{ .Arg .Query.Key.Group }}
+            AND {{ .Ident "resource" }} = {{ .Arg .Query.Key.Resource }}
+            AND {{ .Ident "name" }}     = {{ .Arg .Query.Key.Name }}
+
+            {{ if ne .Query.Key.Namespace "" }}
+                AND {{ .Ident "namespace" }} = {{ .Arg .Query.Key.Namespace }}
+            {{ end }}
```
Comment (chaudyg) [inline pkg/services/store/entity/sqlstash/data/entity_history.sql:23]: Do we need this in the If statement ?

File: pkg/services/store/entity/db/dbimpl/dbEngine.go (hunk 10 above)
Comment (diegommm) [reply pkg/services/store/entity/db/dbimpl/dbEngine.go:56]: The current implementation is broken and won't work, it only works if you use `grafana` / `grafana` (or `grafanatest`) as user and pass. Try adding a space in your password, which broke my local tests.

File: pkg/services/store/entity/sqlstash/data/entity_history.sql (hunk 11 above)
Comment (diegommm) [reply pkg/services/store/entity/sqlstash/data/entity_history.sql:23]: Yes, it is possible to query for entities on all namespaces.

File: pkg/services/store/entity/sqlstash/data/entity_read.sql [hunk 12]
```diff
@@ -73,6 +34,6 @@ SELECT
       {{ end }}
 
       {{ if .SelectForUpdate }}
-          {{ .SelectFor "UPDATE" }}
+          {{ .SelectFor "UPDATE NOWAIT" }}
```
[the remark this answers is not in the document]
Comment (diegommm) [reply pkg/services/store/entity/sqlstash/data/entity_read.sql:37]: This is a change requested in the previous PR: https://github.com/grafana/grafana/pull/87977#discussion_r1608931112

File: pkg/services/store/entity/sqlstash/sql_storage_server_test.go [hunk 13]
```diff
@@ -150,3 +48,18 @@ func setUpTestServer(t *testing.T) entity.EntityStoreServer {
 	require.NoError(t, err)
 	return s
 }
+
+// TODO: remove all the following once the Proposal 1 for Consistent Resource
+// Version is finished.
```
Comment (chaudyg) [inline pkg/services/store/entity/sqlstash/sql_storage_server_test.go:53]: What is this for ?

File: pkg/services/store/entity/sqlstash/data/entity_read.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (chaudyg) [reply pkg/services/store/entity/sqlstash/data/entity_read.sql:37]: Following a discussion: Only here as an optimisation when we Update/Delete a record concurrently. nit

File: pkg/services/store/entity/sqlstash/sql_storage_server.go [hunk 14]
```diff
@@ -191,7 +211,7 @@ func (s *sqlEntityServer) getReadSelect(r FieldSelectRequest) (string, error) {
 	return "SELECT " + strings.Join(quotedFields, ","), nil
 }
 
-func readEntity(rows *sql.Rows, r FieldSelectRequest) (*entity.Entity, error) {
+func oldReadEntity(rows *sql.Rows, r FieldSelectRequest) (*entity.Entity, error) {
```
[the remark this answers is not in the document]
Comment (chaudyg) [reply pkg/services/store/entity/sqlstash/sql_storage_server.go:214]: As discussed. This is used by the read path for now.

File: pkg/services/store/entity/sqlstash/sql_storage_server_test.go [hunk 15]
```diff
@@ -25,113 +25,11 @@ func TestIsHealthy(t *testing.T) {
 	require.NoError(t, err)
 }
 
-func TestCreate(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (diegommm) [reply pkg/services/store/entity/sqlstash/sql_storage_server_test.go:28]: I will provide full test coverage in followup PRs. This test is broken and flaky, and is tightly coupled with the previous code. It fails because of several reasons:

- It doesn't contain enough information in the request to make a successful call, thus it fails in the new code where we check several conditions.
- It doesn't use a proper `context.Context`, containing the expected user information. We check this in the new code and we use this information to populate the `CreatedBy` field.

Overall, fixing this test is not worth it, the next PRs will contain tests with real payloads and realistic failure conditions.

File: pkg/services/store/entity/sqlstash/sql_storage_server_test.go (hunk 13 above)
Comment (diegommm) [reply pkg/services/store/entity/sqlstash/sql_storage_server_test.go:53]: Silences the `unused` linter. These items will be used in followup PRs, and these items are witnesses that they all must be used or otherwise removed.

File: pkg/services/store/entity/db/dbimpl/dbEngine.go (hunk 10 above)
Comment (diegommm) [reply pkg/services/store/entity/db/dbimpl/dbEngine.go:57]: Reduced the items being set to the bare minimum, moved to TODO comments the rest, with documentation references.

File: pkg/services/store/entity/db/dbimpl/dbEngine.go (hunk 9 above)
Comment (diegommm) [reply pkg/services/store/entity/db/dbimpl/dbEngine.go:16]: Reduced the items being set to the bare minimum, moved to TODO comments the rest.

File: pkg/services/store/entity/sqlstash/create.go (hunk 7 above)
Comment (chaudyg) [reply pkg/services/store/entity/sqlstash/create.go:1]: It would be nice to have at least one test for each method : create / update / delete.

File: pkg/services/store/entity/sqlstash/create.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+func (s *sqlEntityServer) Create(ctx context.Context, r *entity.CreateEntityRequest) (*entity.CreateEntityResponse, error) {
+	ctx, span := s.tracer.Start(ctx, "storage_server.Create")
+	defer span.End()
+
+	key, err := entity.ParseKey(r.Entity.Key)
+	if err != nil {
+		return nil, fmt.Errorf("create entity: parse entity key: %w", err)
+	}
+
```
[the remark this answers is not in the document]
Comment (chaudyg) [reply pkg/services/store/entity/sqlstash/create.go:24]: From pairing discussion : this is covered by the insert statement. Let's make sure we have one test for this !

File: pkg/services/store/entity/db/dbimpl/dbimpl.go [hunk 17]
```diff
@@ -45,42 +49,57 @@ func (db *EntityDB) Init() error {
...
+			engine, err = getEngineMySQL(getter, db.tracer)
 			if err != nil {
-				return nil, err
+				return err
 			}
-			_, err = engine.Exec("SELECT 1")
-			if err != nil {
-				return nil, err
+
+			if err = engine.Ping(); err != nil {
```
Comment (chaudyg) [inline pkg/services/store/entity/db/dbimpl/dbimpl.go:97]: I am getting the following error when running locally with mysql :

```
INFO [05-29|11:46:19] Running in alternative execution of Error/NoData mode logger=ngalert.state.manager
[xorm] [info]  2024/05/29 11:46:19.321839 PING DATABASE mysql
Error: ✗ initialize Entity Server: Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '' at line 1
[Bra] 05-29 11:46:19 [ WARN] Fail to execute command: ./bin/grafana [server -packaging=dev cfg:app_mode=development] - exit status 1
```

File: pkg/services/store/entity/sqlstash/delete.go [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+	err = s.sqlDB.WithTx(ctx, ReadCommitted, func(ctx context.Context, tx db.Tx) error {
+		// Pre-locking: get the latest version of the entity
+		previous, err := readEntity(ctx, tx, s.sqlDialect, key, r.PreviousVersion, true, false)
+		if errors.Is(err, ErrNotFound) {
+			ret.Status = entity.DeleteEntityResponse_NOTFOUND
+			return nil
+		}
+		if err != nil {
+			return err
```
Comment (chaudyg) [inline pkg/services/store/entity/sqlstash/delete.go:39]: Are we missing rsp.Status = entity.DeleteEntityResponse_ERROR ?

File: pkg/services/store/entity/sqlstash/folder_support.go [hunk 19]
```diff
@@ -23,64 +26,109 @@ type folderInfo struct {
...
 
-	root, lost, err := buildFolderTree(all)
-	if err != nil {
-		return err
+	if err := rows.Close(); err != nil {
+		return fmt.Errorf("close rows after listing folder items in namespace %q: %w", namespace, err)
 	}
 
-	err = insertFolderInfo(ctx, tx, namespace, root, false)
+	root, lost, err := buildFolderTree(itemList)
```
Comment (chaudyg) [inline pkg/services/store/entity/sqlstash/folder_support.go:107]: Mabye another FIXME.

Is this working with concurrent executions ? Two concurrent transactions might build a version of the folder tree that will not include created/deleted/updated items from the concurrent transaction. Is this assumption correct ?

**Review by chaudyg — COMMENTED** [review]
Overall looks good. Thanks for doing this.

My only ask would be to add a few tests for the Create/Update/Delete methods.

**Comment by diegommm** [comment]
> Overall looks good. Thanks for doing this.
> 
> My only ask would be to add a few tests for the Create/Update/Delete methods.

Thank you! I promise the next PR will provide the necessary tests. I really want tests to be reviewed in a separate PR so that there is some reviewing dedicated exclusively to tests to make sure they are relevant and correct as well.

**Comment by diegommm** [comment]
Rebased and force-pushed.

**Review by chaudyg — APPROVED** [review]
